New Ownership Property: Notes

ewhac opened the pull request:
The owner of a submissive may now "pin a note" to them ("If found, return to...", "Do not remove $(ITEM_LIST)...," "Is being disciplined for...," etc.).  The note is free-form (compressed) text stored in the Ownership struct, and is displayed in the user's OnlineProfile pane.

Create a new Ownership property: `Notes`.  If the updated note has zero length, the property is set to null.

Create a new Message request type: `ServerAccountOwnershipUpdateNotesRequest`.

ewhac committed:
New Ownership Property: Notes

The owner of a submissive may now "pin a note" to them ("If found,
return to...", "Do not remove $(ITEM_LIST)...," "Is being disciplined
for...," etc.).  The note is free-form (compressed) text stored in the
Ownership struct, and is displayed in the user's OnlineProfile pane.

Create a new Ownership property: `Notes`.  If the updated note has zero
length, the property is set to null.

Create a new Message request type: `ServerAccountOwnershipUpdateNotesRequest`.

ewhac commented:
See https://gitgud.io/BondageProjects/Bondage-College/-/merge_requests/5968 for the client-side merge request.

zorgjbe commented on app.js:
```diff
@@ -2356,19 +2356,41 @@ function AccountOwnership(data, socket) {
+		    &&  TargetAcc.Ownership != null
+		    &&  TargetAcc.Ownership.MemberNumber == Acc.MemberNumber)
+		{
+			if (data.Notes != null  &&  typeof data.Notes === "string"  &&  data.Notes.length > 0) {
```
null check is unnecessary here as there's already a `typeof` check for the same prop.

zorgjbe commented on app.js:
```diff
@@ -2356,19 +2356,41 @@ function AccountOwnership(data, socket) {
+		// The dominant is setting/updating public notes on their submissive.
+		if (    data.Action === "UpdateNotes"
+		    &&  TargetAcc.Owner != null
+		    &&  TargetAcc.Owner !== ""
```
I'm curious about that one. Ideally, the client is just using it to set up NPC ownerships, and nothing else so I'm wondering why it's being checked as well.

ewhac replied:
`TargetAcc.Owner` isn't set until the member is fully owned (i.e. completed the trial period and accepted the collar).

zorgjbe replied:
Oh, so this is a fully owned check? Why not look at `Stage` then?

zorgjbe commented on app.js:
```diff
@@ -2356,19 +2356,41 @@ function AccountOwnership(data, socket) {
+			Database.collection(AccountCollection).updateOne ({ AccountName: TargetAcc.AccountName },
+			                                                  { $set: O },
+			                                                  function (err, _res) { if (err) throw err; });
+			socket.emit ("AccountOwnership", O);
```
Technically those two lines go inside the `updateOne` callback — ensures the db is done with the update before telling about it.

ewhac replied:
Good point.

zorgjbe commented on app.js:
```diff
@@ -2356,19 +2356,41 @@ function AccountOwnership(data, socket) {
 			TargetAcc.Ownership = null;
 			let O = { Ownership: TargetAcc.Ownership, Owner: TargetAcc.Owner };
-			Database.collection(AccountCollection).updateOne({ AccountName : TargetAcc.AccountName }, { $set: O }, function(err, res) { if (err) throw err; });
+			Database.collection(AccountCollection).updateOne({ AccountName : TargetAcc.AccountName }, { $set: O }, function(err, _res) { if (err) throw err; });
```
I'd just ignore those, they're just diff noise here.

ewhac committed:
Code review patches requested by @zorgjbe

ewhac committed:
Code review changes

When determining fully owned-ness, switch from looking at
`TargetAcc.Owner` to `TargetAcc.Ownership.Stage`.

If the notes string is missing or empty, set the `Notes` field to
`undefined` instead of `null`.  (The effect in the DB appears to be the
same).

zorgjbe reviewed (approved):
Looks good to me. Just a couple smaller notes:
- the style 😁 
- 10000 sounds like a bit too large of a description for that. I'd limit it to something smaller like 1-2000 or so? Maybe even less… 'cause the larger it is, the more data has to be carried around and synced all over the place to every room participant.

ewhac committed:
Coding style changes

No functional changes.  Alter coding style of changes to more closely
conform to the extant style.

ewhac commented:
> * 10000 sounds like a bit too large of a description for that. I'd limit it to something smaller like 1-2000 or so?

How about 4096?  (You just *know* some people are going to try write a novella in these things, right?)

EDIT: ...mmhhhrrr...  Okay, 4000.  Not everyone's a computer geek.

ewhac committed:
Reduce max owner's notes length from 10000 to 4000

ewhac committed:
Fix owner's notes update back to client

`socket.emit()` is for the account associated with the current
connection/login.  If we want to send Ownership changes to a different
user, we have to send it to their socket.  Hence, when updating owner's
notes, send the Ownership update via `TargetAcc.Socket.emit()`.

Direct the chat room sync toward the `TargetAcc` character as well when
calling `ChatRoomSyncCharacter()`.

bananarama92 commented:
> How about 4096? (You just know some people are going to try write a novella in these things, right?)

Is this a use case that should even be accommodated? IMO it's _very_ fair to just force them to use normal online descriptions at that point if someone feels inclined to write a novel.

ewhac commented:
> IMO it's _very_ fair to just force them to use normal online descriptions at that point if someone feels inclined to write a novel.

Except that owners can't alter their subs' online profile (without using BCX).

In any case, this limit can now be altered by changing a single line in the code, so if 4000 chars is at some point in the future deemed too burdensome, it can be reduced.
